File: lib/206.reverse_linked_list.dart
import 'other/list_node.dart';

void main() {
  final list1 = reverseList(ListNode(1)
    ..next = (ListNode(2)..next = (ListNode(3)..next = (ListNode(4)..next = ListNode(5)))));
  list1?.printList();

  final list2 = reverseList(ListNode(1)..next = ListNode(2));
  list2?.printList();
}

ListNode? reverseList(ListNode? head) {
  var root = ListNode(0);

  while (head!=null) {
    final preNext = root.next;
    root.next = head;
    head = head.next;
    root.next?.next = preNext;
  }

  return root.next;
}

File: src/206.reverse_linked_list.kt
import other.ListNode
import other.printList

fun main() {
    val list1 = reverseList(ListNode(1).apply {
        next = ListNode(2).apply {
            next = ListNode(3).apply {
                next = ListNode(4).apply {
                    next = ListNode(5)
                }
            }
        }
    })

    list1?.printList()

    val list2 = reverseList(ListNode(1).apply {
        next = ListNode(2)
    })

    list2?.printList()
}

fun reverseList(head: ListNode?): ListNode? {
    val root = ListNode(0)
    var current = head

    while (current!=null) {
        val preNext = root.next
        root.next = current
        current = current.next
        root.next?.next = preNext
    }

    return root.next
}